README.md
# svelte-sequence

Custom stores to compose tweened motion sequences over multiple steps.

- `tweenedSequence` to compose tweened motion sequences over multiple steps
- `stopwatch` to create an instance of a stopwatch
- `timer` to create an instance of a timer

**Installation:**
`npm i -D svelte-sequence`

Example on StackBlitz: https://stackblitz.com/edit/example-svelte-sequence

# tweenedSequence

**Methods:**

- `setStep(step, options?)`
- `updateSequence((currentSequence) => newSequence, options?)`
- `nextStep(options?)`
- `previousStep(options?)`

**Stores:**
- `value: Tweened<T>`
- `step: Writable<string | number>`


## **Usage**

Imagine you have an object on screen. You want this object to move to specific points as a user clicks through some awesome tutorial you made. Your tutorial has four steps.

1. Create a `tweenedSequence` store. Provide it with your sequence of values at each step:

```js
const sequence = {
	step_one: { x: 0, y: 0 },
	step_two: { x: 10, y: 5 },
	step_three: { x: 15, y: 10 },
	step_four: { x: 50, y: 25 }
};

const myPosition = tweenedSequence(sequence, { initialStep: 'step_one' });
```

2. The `myPosition` store you created will emit the `{ x, y }` values based on the step you set. Use your sequence keys to update the step and consume the new `{ x, y }` values:

```js
// will tween from current value to 'step_two' value
myPosition.setStep('step_two');
```

And somewhere in your template you could:

```svelte
<div class="dot" style:left="{$myPosition['x']}px" style:top="{$myPosition['y']}px" />
```

## **tweenedSequence Args**

`tweenedSequence(sequence, options)`

### **sequence**

A `tweenedSequence` accepts a sequence as objects or arrays.

**Named Sequence (objects)**

Each key in your named sequence corresponds to a step. You can call `setStep(name)` for any key in the provided sequence.

If you call `setStep()` with a name that is not present in a sequence, nothing will happen. This allows for flexible step coordination between objects without errors getting thrown or unwanted movement.

Named Sequence _values_ can be anything a regular tweened store accepts (number, array, object). The values must be the same "shape" across steps.

```js
const namedSequence_number = {
	step_one: 0,
	step_two: 8
	// ...
};

const namedSequence_array = {
	step_one: [0, 0],
	step_two: [8, 14]
	// ...
};

const namedSequence_object = {
	step_one: { x: 0, y: 0 },
	step_two: { x: 8, y: 14 }
	// ...
};
```

**Indexed Sequence (arrays)**

Each index in your indexed sequence corresponds to a step. You can call `setStep(index)` for any index in the provided sequence.

If you call `setStep()` with a fractional value that exists in the sequence, you'll receive the value at that fractional point in the tween based on your provided easing. This allows for "scrubbing" through your motion.

For example, `setStep(1.5)` will emit the value halfway between index 1 and index 2 in the sequence you provided.

Indexed Sequence _values_ can be anything a regular tweened store accepts (number, array, object). The values must be the same "shape" across steps.

```js
const indexedSequence_number = [0, 8];

const indexedSequence_array = [
	[0, 0],
	[8, 14]
	// ...
];

const indexedSequence_object = [
	{ x: 0, y: 0 },
	{ x: 8, y: 14 }
	// ...
];
```

**Step Binding**

You can also bind directly to `step` which is returned from `tweenedSequence`. Combine this with a range slider for easy scrubbing:

```js
const myPosition = tweenedSequence(indexedSequence_array);
const { step } = myPosition;
```

And in your template:

```svelte
<label>Scrubber: {$step.toFixed(2)}
	<input type="range" bind:value={$step} step="0.01" min="0" max={indexedSequence_array - 1} />
</label>
```

### **options**

Similar to regular tweened stores, the `tweenedSequence` options accept the following optional properties:

```js
const options = {
	duration: 400, // number, ms
	easing: linear, // any easing fn from svelte/easing
	delay: 0, // number, ms
	initialStep: 0 // the initial step of your sequence (name or index)
};
```

## **updateSequence()**

If your sequence changes for some reason, you may need to update parts of it.

For example an object moves to four positions, but the third position is determined by user click.

```js
function handleClick(event) {
	mySequence.updateSequence((current) => {
		// index 2 is the third item in the sequence
		current[2] = [event.clientX, event.clientY];
		return current;
	});
}
```

And voila! Your sequence is updated and your call to `setStep(2)` will tween to your new tuple position. This works for any sequence type or value type, as long as the returned sequence is the same shape as the original.

## **Dynamic Options**

Also similar to a tweened sotre, you can update your options (like delay, easing, or duration) with an optional second argument to any of your `setStep`, `previousStep`, `nextStep`, or `updateSequence` calls.

For example:
```js
function elasticNext(){
	mySequence.nextStep({ easing: elasticOut, duration: 3000 });
}

function cubicPrevious(){
	mySequence.previousStep({ easing: cubicInOut, duration: 800 });
}

```
# stopwatch

A stopwatch store that emits time elapsed in ms.

**Methods:**

- `start()`
- `pause()`
- `reset()`
- `addLap()`

**properties:**

- `laps: Writable<Lap[]>` - a store with an array of registered laps

**Usage**
Create a new stopwatch:

```js
const firstStopwatch = stopwatch();
```

Invoke any of the methods:

```js
firstStopwatch.start();
firstStopwatch.pause();
firstStopwatch.reset();
firstStopwatch.addLap();
```

Extract and read the laps as needed:

```svelte
<script>
//...
const { laps } = firstStopwatch;

</script>

{#each $laps as lap (lap.lapNumber)}
  <LapDisplay {lap}>
{/each}
```

## Timer

A stopwatch store that emits time elapsed in ms.

**method:**

- `start()`
- `pause()`
- `reset()`
- `add(time, timeUnit = 'ms')`

**Usage**
Create a new timer:
```js
// Creates a timer of 10000 ms
const firstTimer = timer(10000);

// Creates a timer of 1.5 hours
const secondTimer = timer(1.5, 'h');
```
**NOTE:** timers will always emit their value in ms, regardless of how you initialize them. There are a handful of time utils exported to help with common time parsing.

Invoke any of the methods:

```js
firstTimer.start();
firstTimer.pause();
firstTimer.reset();

// adds 60 seconds to the timer, useful for "snooze" scenarios
firstTimer.add(60, 's');
```

## Time Utils
```ts
// Converts ms to Seconds
function getSeconds(ms: number) {
	return Math.floor(ms / 1000);
}
// Converts ms to Minutes
function getMinutes(ms: number) {
	return Math.floor(ms / minMS);
}
// Converts ms to Hours
function getHours(ms: number) {
	return Math.floor(ms / hourMS);
}
// Converts ms to Days
function getDays(ms: number) {
	return Math.floor(ms / dayMS);
}
// Converts ms to Weeks
function getWeeks(ms: number) {
	return Math.floor(ms / weekMS);
}

// Converts ms (t) into constituent units of time, returned as an array
function formatTime(t: number, includeZero: boolean = false) {
	const { wk, d, h, m, s } = extractUnits(t);
	const timeList = [
		{ time: wk, unit: wk === 1 ? 'week' : 'weeks', shortUnit: 'wk' },
		{ time: d, unit: d === 1 ? 'day' : 'days', shortUnit: 'd' },
		{ time: h, unit: h === 1 ? 'hour' : 'hours', shortUnit: 'hr' },
		{ time: m, unit: m === 1 ? 'minute' : 'minutes', shortUnit: 'min' },
		{ time: s / 1000, unit: s === 1 ? 'second' : 'seconds', shortUnit: 's' }
	];
	if (includeZero) {
		return timeList;
	} else {
		return timeList.filter((tData) => tData.time > 0);
	}
}
```

## Types

```ts
interface TweenedSequence<T> extends Readable<T> {
	value: Tweened<T>,
	step: Writable<string | number>;
	setStep: (step: string | number, options?: TweenedSequenceOptions) => void;
	nextStep: (options?: TweenedSequenceOptions) => void;
	previousStep: (options?: TweenedSequenceOptions) => void;
	// PRs welcome :D
	updateSequence: (fn: (currentSequence: any) => any, options?: TweenedSequenceOptions) => void;
}

interface TweenedSequenceOptions {
	duration?: number;
	delay?: number;
	easing?: (t: number) => number;
	initialStep?: number | string;
}

interface Lap {
	lapNumber: number;
	lapTime: number;
	totalTime: number;
	delta: number;
}

type TimeUnit = 'ms' | 's' | 'm' | 'h' | 'd' | 'w';
```

## Roadmap

- Better type support
- Add function metadata for better dx


CHANGELOG.md
# svelte-sequence

## 0.4.0

### Minor Changes

- 648229c: Add optional sequence options to all methods

## 0.3.1

### Patch Changes

- 3da2d51: fix bad error message for invalid initialValue of indexed sequence

## 0.3.1

### Patch Changes

- ad593c5: tweenedSequence returns value and step stores

## 0.3.0

### Minor Changes

- 428a7ac: Add updateSequence method, marginally improve types

## 0.2.0

### Minor Changes

- 2b0dde8: add previousStep() and nextStep() methods, fix time store memory leaks

## 0.1.0

### Minor Changes

- 26c30f2: Add timer store, stopwatch store, basic time formatting utils

## 0.0.2

### Patch Changes

- 964a0ae: initial version


src/lib/timer.ts
import { writable } from 'svelte/store';
import type { TimeUnit } from './models';
import { convertTimeAndUnitToMS } from './timeUtils';

function createTimer(t: number, u: TimeUnit) {
	let intervalID: ReturnType<typeof setInterval> | null;
	const clear = () => {
		if (intervalID) {
			clearInterval(intervalID);
			intervalID = null;
		}
	} 
	const originalTime = convertTimeAndUnitToMS(t, u);
	const { subscribe, set, update } = writable(originalTime, () => () => {
		if (intervalID) clear();
	});

	return {
		subscribe,
		start: () => {
			if (intervalID) return;
			intervalID = setInterval(() => {
				update((currentTime) => currentTime - 10);
			}, 10);
		},
		pause: () => {
			if (intervalID) clear();
		},
		reset: () => {
			if (intervalID) clear();
			set(originalTime);
		},
		add: (time: number, unit: TimeUnit = 'ms') => {
			const duration = convertTimeAndUnitToMS(time, unit);
			update((currentTime) => currentTime + duration);
		}
	};
}

export const timer = (time: number, unit: TimeUnit = 'ms') => createTimer(time, unit);


src/lib/models.ts
import type { Tweened } from 'svelte/motion';
import type { Readable, Writable } from 'svelte/store';
export type TimeUnit = 'ms' | 's' | 'm' | 'h' | 'd' | 'w';

export type SequenceType = 'NAMED' | 'INDEXED';
// export type TweenedValue =
// 	| { [key: string]: number }
// 	| number
// 	| number[]
// 	| [number, number]
// 	| [number, number, number]
// 	| [number, number, number, number];

export type NamedSequence<T> = {
	[key: string]: T;
};
export type IndexedSequence<T> = T[];

export type Sequence<T> = NamedSequence<T> | IndexedSequence<T>;

// export interface IndexedTweenedSequence<T> extends Readable<T> {
	// setStep: (step: number) => void;
	// nextStep: () => void;
	// previousStep: () => void;
	// step: Readable<string | number>;
	// step: string | number;
// }

// type NameOrId<T extends number | string> = T extends number
//   ? IdLabel
//   : NameLabel;
// export interface NamedTweenedSequence<T> extends Readable<T> {
	// setStep: (step: string) => void;
	// nextStep: () => void;
	// previousStep: () => void;
	// step: Readable<string | number>;
	// step: string | number;
// }

// type Updater<T> = (target_value: T) => T;

export interface TweenedSequence<T> extends Readable<T> {
		setStep: (step: string | number, options?: TweenedSequenceOptions) => void;
		nextStep: (options?: TweenedSequenceOptions) => void;
		previousStep: (options?: TweenedSequenceOptions) => void;
		value: Tweened<T>,
		step: Writable<string | number>;
		// TODO: figure out how to type this update fn.
		/**
		 * 
		 * @param fn Update function that receives the current sequence as the first arg 
		 * and returns the new sequence.
		 * 
		 * NOTE: the new sequence must be the same "shape" as the original sequence.
		 * 
		 */
		updateSequence: (fn: (currentSequence: any) => any, options?: TweenedSequenceOptions) => void;
		// updateSequence: (updater: (s: Sequence<T>) => Sequence<T>) => void

}

export interface TweenedSequenceOptions {
	duration?: number;
	delay?: number;
	easing?: (t: number) => number;
	initialStep?: number | string;
}

export interface Lap {
	lapNumber: number,
	lapTime: number,
	totalTime: number,
	delta: number
}

// FOR INTERNAL USE - DO NOT EXPORT TO PACKAGE
export type Vector2 = {
	x: number;
	y: number;
	clone: () => void;
	setX: (n: number) => void;
	setY: (n: number) => void;
	setZ: (n: number) => void;
};
export type Vector3 = {
	x: number;
	y: number;
	z: number;
	clone: () => Vector3;
	setX: (n: number) => Vector3;
	setY: (n: number) => Vector3;
	setZ: (n: number) => Vector3;
};


src/lib/stopwatch.ts
import { get, writable } from 'svelte/store';
import type { Lap } from './models';

function createStopwatch(startAt: number) {
	let intervalID: ReturnType<typeof setInterval> | null;
	const clear = () => {
		if (intervalID) {
			clearInterval(intervalID);
			intervalID = null;
		}
	} 

	const originalTime = startAt;
	const stopwatchStore = writable(startAt, () => () => {
		if (intervalID) clear();
	});
	const { subscribe, set, update } = stopwatchStore;
	const laps = writable<Lap[]>([]);
	return {
		subscribe,
		start: () => {
			if (intervalID) return;

			intervalID = setInterval(() => {
				update((currentTime) => currentTime + 10);
			}, 10);
		},
		pause: () => {
			if (intervalID) clear();
		},
		reset: () => {
			if (intervalID) clear();
			set(originalTime);
			laps.set([]);
		},
		laps,
		addLap: () => {
			laps.update((currentLaps) => {
				const prevLap = currentLaps[0];
				const totalTime = get(stopwatchStore);
				const prevTotalTime = prevLap?.totalTime ?? 0;
				const lapTime = totalTime - prevTotalTime;
				const lapDelta = prevLap ? lapTime - (prevLap.lapTime) : 0;
				const lapNumber = currentLaps.length + 1;
				return [
					{
						lapNumber,
            lapTime,
						totalTime,
						delta: lapDelta
					},
					...currentLaps
				];
			});
		}
	};
}

export const stopwatch = (startTime = 0) => createStopwatch(startTime);


src/lib/tweenedSequence.ts
import { tweened } from 'svelte/motion';
import { linear } from 'svelte/easing';
import type {
	IndexedSequence,
	// IndexedTweenedSequence,
	NamedSequence,
	// NamedTweenedSequence,
	Sequence,
	SequenceType,
	TweenedSequence,
	TweenedSequenceOptions,
	Vector3
} from './models';
import { writable } from 'svelte/store';
import { onDestroy } from 'svelte';

const defaultOptions: TweenedSequenceOptions = {
	duration: 1000,
	easing: linear,
	delay: 0
};

function calculateNextValueFromFract<T>(a: T, b: T, factor: number): T {
	if (typeof a === 'number' && typeof b === 'number') {
		// handle numbers
		return ((b - a) * factor + a) as T;
	}

	if (Array.isArray(a) && Array.isArray(b)) {
		// handle arrays
		const temp = [...a];
		a.forEach((val, idx) => {
			temp[idx] = (b[idx] - val) * factor + val;
		});
		return temp as T;
	}

	// TODO: See if this even works to handle threejs vector3
	// @ts-ignore
	if (a.isVector3 && b.isVector3) {
		// handle threejs vector3
		const v3 = (a as Vector3)
			.clone()
			.setX(
				(((b as Vector3)['x'] as number) - ((a as Vector3)['x'] as number)) * factor +
					((a as Vector3)['x'] as number)
			)
			.setY(
				(((b as Vector3)['y'] as number) - ((a as Vector3)['y'] as number)) * factor +
					((a as Vector3)['y'] as number)
			)
			.setZ(
				(((b as Vector3)['z'] as number) - ((a as Vector3)['z'] as number)) * factor +
					((a as Vector3)['z'] as number)
			);
		return v3 as T;
	}

	if (!!a && !!b && typeof a === 'object' && typeof b === 'object') {
		// handle objects
		const temp = {};
		Object.keys(a).forEach((k) => {
			// TODO
			// @ts-ignore
			temp[k] = ((b[k] as number) - (a[k] as number)) * factor + (a[k] as number);
		});
		return temp as T;
	}

	return b;
}

function updateOptions(
	currentOptions: TweenedSequenceOptions,
	newOptions: TweenedSequenceOptions | undefined
): TweenedSequenceOptions {
	if (!newOptions) return currentOptions;

	let duration = newOptions.duration ?? currentOptions.duration;
	let easing = newOptions.easing ?? currentOptions.easing;
	let delay = newOptions.delay ?? currentOptions.delay;
	let initialStep = newOptions.initialStep ?? currentOptions.initialStep;
	return { duration, easing, delay, initialStep };
}

function createIndexedSequence<T>(
	_sequence: IndexedSequence<T>,
	options: TweenedSequenceOptions
): TweenedSequence<T> {
	let sequence = _sequence;
	const seqType: SequenceType = 'INDEXED';
	let { duration, delay, easing, initialStep: _iStep } = updateOptions(defaultOptions, options);
	let initialStep = (_iStep as number) ?? 0;

	// Validate initial step
	if (initialStep < 0 || initialStep > sequence.length) {
		console.error(
			`Step "${initialStep}" is an invalid initialStep for your sequence of length ${
				sequence.length
			}. A valid step satisfies: 0 <= step <= ${sequence.length - 1}`
		);
	}

	function syncOptions(newOptions: TweenedSequenceOptions | undefined) {
		const updatedOptions = updateOptions({ duration, easing, delay, initialStep }, newOptions);
		duration = updatedOptions.duration;
		easing = updatedOptions.easing;
		delay = updatedOptions.delay;
		initialStep = updatedOptions.initialStep as number;
	}

	const step = writable(initialStep);
	const value = tweened<T>(sequence[initialStep], { duration, easing, delay });
	const { subscribe, set, update } = value;
	let previousStep: number = initialStep;

	// lil fractional helper
	const fract = (num: number) => num % 1.0;

	const setStep = (step: string | number) => {
		if (typeof step !== 'number') {
			console.error(
				`Step "${step}" is not assignable to an indexed tweenedSequence store. "${step}" is not a number.`
			);
			return;
		}

		// NOTE: Ignore invalid steps (alternative would be to round, but seems weird)
		if (step > sequence.length - 1 || step < 0) {
			return;
		}

		let nextDuration;
		let nextDelay;

		let partial = fract(step);
		const previousValue: T = sequence[Math.floor(step)] ?? sequence[0];
		const nextValue: T = sequence[Math.ceil(step)] ?? sequence[1];
		const nextFactor = easing!(partial);
		const stepDiff = Math.abs(previousStep - step);
		nextDuration = stepDiff * duration! < duration! ? stepDiff * duration! : duration;
		// TODO: consider delay behavior (when to apply, when not to)
		// maybe as arg to setStep so user can decide?
		nextDelay = stepDiff * delay! < delay! ? stepDiff * delay! : delay;
		const nextVal: T = calculateNextValueFromFract<T>(
			previousValue, // from
			nextValue, // to
			nextFactor
		);
		previousStep = step;
		return set(nextVal, {
			duration: nextDuration ?? duration,
			delay: nextDelay ?? delay,
			easing
		});
	};

	const unsubscribe = step.subscribe(setStep);

	onDestroy(unsubscribe);

	return {
		value,
		step,
		subscribe,
		nextStep: (newOptions) => {
			const baseNext = Math.floor(previousStep + 1);
			// For sequence array, if next idx is too high, use 0
			const actualNext = baseNext > sequence.length - 1 ? 0 : baseNext;
			syncOptions(newOptions);
			return step.set(actualNext);
		},
		previousStep: (newOptions) => {
			const basePrev = Math.ceil(previousStep - 1);
			// For sequence array, if next idx is too high, use 0
			const actualPrev = basePrev < 0 ? sequence.length - 1 : basePrev;
			syncOptions(newOptions);
			return step.set(actualPrev);
		},
		setStep: (s, newOptions) => {
			syncOptions(newOptions);
			step.set(s as number);
		},
		updateSequence: (fn, newOptions) => {
			sequence = fn(sequence) as IndexedSequence<T>;
			syncOptions(newOptions);
			return setStep(previousStep);
		}
	};
}

function createNamedSequence<T>(
	_sequence: NamedSequence<T>,
	options: TweenedSequenceOptions
): TweenedSequence<T> {
	let sequence = _sequence;
	const seqType: SequenceType = 'NAMED';
	let { duration, delay, easing, initialStep: _iStep } = updateOptions(defaultOptions, options);
	let initialStep = (_iStep as string) ?? Object.keys(sequence)[0];

	// Validate initial step
	if (sequence[initialStep] === null || sequence[initialStep] === undefined) {
		console.error(
			`Step "${initialStep}" is an invalid initialStep for your named sequence. A valid step will match one of your sequence keys: ${Object.keys(
				sequence
			).join(', ')}`
		);
	}

	function syncOptions(newOptions: TweenedSequenceOptions | undefined) {
		const updatedOptions = updateOptions({ duration, easing, delay, initialStep }, newOptions);
		duration = updatedOptions.duration;
		easing = updatedOptions.easing;
		delay = updatedOptions.delay;
		initialStep = updatedOptions.initialStep as string;
	}

	const step = writable(initialStep);
	const value = tweened<T>(sequence[initialStep], { duration, easing, delay });
	const { subscribe, set, update } = value;
	let previousStep: string = initialStep;

	function setStep(step: string | number) {
		if (typeof step === 'number' || !sequence.hasOwnProperty(step)) {
			return;
		}

		previousStep = step;
		return set(sequence[step] ?? sequence[initialStep], {
			duration,
			delay,
			easing
		});
	}

	const unsubscribe = step.subscribe(setStep);
	onDestroy(unsubscribe);

	return {
		step,
		value,
		subscribe,
		nextStep: (newOptions) => {
			const keys = Object.keys(sequence);
			const currentIdx = keys.indexOf(previousStep);
			const baseNextKeyIdx = currentIdx + 1;
			const actualNext = keys[baseNextKeyIdx] ?? keys[0];
			syncOptions(newOptions);
			return step.set(actualNext);
		},
		previousStep: (newOptions) => {
			const keys = Object.keys(sequence);
			const currentIdx = keys.indexOf(previousStep as string);
			const basePrevKeyIdx = currentIdx - 1;
			const actualPrev = keys[basePrevKeyIdx] ?? keys[keys.length - 1];
			syncOptions(newOptions);
			return step.set(actualPrev);
		},
		setStep: (s, newOptions) => {
			syncOptions(newOptions);
			step.set(s as string);
		},
		updateSequence: (fn, newOptions) => {
			sequence = fn(sequence) as NamedSequence<T>;
			syncOptions(newOptions);
			return setStep(previousStep);
		}
	};
}

export function tweenedSequence<T>(sequence: Sequence<T>, options: TweenedSequenceOptions = {}) {
	const seqType: SequenceType = Array.isArray(sequence) ? 'INDEXED' : 'NAMED';
	if (seqType === 'INDEXED') {
		return createIndexedSequence<T>(sequence as IndexedSequence<T>, options);
	} else {
		// (seqType === 'NAMED')
		return createNamedSequence<T>(sequence as NamedSequence<T>, options);
	}
}


src/lib/timeUtils.ts
import type { TimeUnit } from "./models";

export const minMS = 60_000;
export const hourMS = minMS * 60;
export const dayMS = hourMS * 24;
export const weekMS = dayMS * 7;
export function getSeconds(ms: number) {
	return Math.floor(ms / 1000);
}
export function getMinutes(ms: number) {
	return Math.floor(ms / minMS);
}
export function getHours(ms: number) {
	return Math.floor(ms / hourMS);
}
export function getDays(ms: number) {
	return Math.floor(ms / dayMS);
}
export function getWeeks(ms: number) {
	return Math.floor(ms / weekMS);
}

function extractUnits(t: number) {
	const wk = Math.floor(t / weekMS);
	const wkms = wk * weekMS;
  const d = Math.floor((t - wkms) / dayMS);
  const dms = d * dayMS;
	const h = Math.floor((t - (wkms + dms)) / hourMS);
	const hms = h * hourMS;
	const m = Math.floor((t - (wkms + dms + hms)) / minMS);
	const mms = m * minMS;
	const s = (t - (wkms + dms + hms + mms)) % minMS;
	return {
		wk,
    d,
		h,
		m,
		s
	};
}

export function formatTime(t: number, includeZero: boolean = false) {
	const { wk, d, h, m, s } = extractUnits(t);
	const timeList = [
		{ time: wk, unit: wk === 1 ? 'week' : 'weeks', shortUnit: 'wk' },
		{ time: d, unit: d === 1 ? 'day' : 'days', shortUnit: 'd' },
		{ time: h, unit: h === 1 ? 'hour' : 'hours', shortUnit: 'hr' },
		{ time: m, unit: m === 1 ? 'minute' : 'minutes', shortUnit: 'min' },
		{ time: s / 1000, unit: s === 1 ? 'second' : 'seconds', shortUnit: 's' }
	];
	if (includeZero) {
		return timeList;
	} else {
		return timeList.filter((tData) => tData.time > 0);
	}
}

export function convertTimeAndUnitToMS(t: number, u: TimeUnit){
  switch(u){
    case "ms":
      return t;
    case 's':
      return t * 1000;
    case 'm':
      return t * minMS;
    case 'h':
      return t * hourMS;
    case 'd':
      return t * dayMS;
    case 'w':
      return t * weekMS;
    default:
      console.warn("Unrecognized unit in timer:", u);
      return t;
  }
}